### PR TITLE
Preserve substs in opaques recorded in typeck results

### PR DESCRIPTION
**Pull request by compiler-errors**
This means that we now prepopulate MIR with opaques with the right substs.

The first commit is a hack that I think we discussed, having to do with `DefiningAnchor::Bubble` basically being equivalent to `DefiningAnchor::Error` in the new solver, so having to use `DefiningAnchor::Bind` instead, lol.

r? @lcnr

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs [hunk 1]
```diff
@@ -208,27 +223,51 @@ pub(super) fn find_opaque_ty_constraints_for_rpit(
...
+                if let Some(prev) = &mut opaque_ty {
+                    if concrete_type.ty != prev.ty && !(concrete_type, prev.ty).references_error() {
+                        prev.report_mismatch(&concrete_type, def_id, tcx).emit();
+                    }
+                } else {
+                    opaque_ty = Some(concrete_type);
+                }
+            }
+
+            opaque_ty.map_or_else(
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs:257]: can this be an `if let` instead?

**Review by lcnr — COMMENTED** [review]
r=me after nits

there's still the FIXME that the substs are wrong when calling `register_predefined_opaques_in_new_solver`, please remove that

File: compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs [hunk 2]
```diff
@@ -190,109 +205,58 @@ impl<'tcx> intravisit::Visitor<'tcx> for TaitConstraintLocator<'tcx> {
...
 ) -> Ty<'_> {
-    let concrete = tcx.mir_borrowck(owner_def_id).concrete_opaque_types.get(&def_id).copied();
-
-    if let Some(concrete) = concrete {
-        let scope = tcx.hir().local_def_id_to_hir_id(owner_def_id);
-        debug!(?scope);
-        let mut locator = RpitConstraintChecker { def_id, tcx, found: concrete };
+    let tables = tcx.typeck(owner_def_id);
 
-        match tcx.hir().get(scope) {
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs:205]: Revert this

**Comment by oli-obk** [comment]
I wanna review this on Monday.

**Comment by compiler-errors** [comment]
Addressed outstanding revert comment.

@bors r=lcnr

**Pull request merged**